### PR TITLE
fix: recover from truncated JSON in tool call arguments

### LINKED ISSUE #4240: Testing judge often fails to parse arguments
### Bug Description

When using the testing assertions with judge, we started seeing more and more cases of

```
services/agent_worker/src/agent/testing/run_story.py:501: in _assert_event
    await chat_assert.judge(llm_judge, intent=story_event.judge)
services/agent_worker/.venv/lib/python3.12/site-packages/opentelemetry/util/_decorator.py:71: in async_wrapper
    return await func(*args, **kwargs)  # type: ignore
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
services/agent_worker/.venv/lib/python3.12/site-packages/livekit/agents/llm/utils.py:361: in prepare_function_arguments
    args_dict = from_json(json_arguments)
                ^^^^^^^^^^^^^^^^^^^^^^^^^
E   ValueError: EOF while parsing a string at line 1 column 66
```

which seems to be related to the judge not producing function arguments, which livekit expects.

We are not sure if we can do anything about this / there needs to be some specific judge prompt tuning.

We never had this issue 2 months ago or so, and this started to be roughly ~10% (vibes based estimate) of our failure cases at this point.

### Expected Behavior

The judge should not fail with false arguments. We know LLMs can be tricky though.

### Reproduction Steps

```bash
A little hard to reproduce at all. Any case where the judge is used with different intents.

We usually phrase intents like this (examples that lead to this failure):

* "Asks reason to transfer"
* "Lets the user know about the one open account and offers to make a payment."
```

### Operating System

Linux (github actions)

### Models Used

_No response_

### Package Versions

```bash
❯ uv pip list
Package                                  Version         Editable project location
---------------------------------------- --------------- -------------------------------------------------------------------------
aiofiles                                 24.1.0
aiohappyeyeballs                         2.6.1
aiohttp                                  3.12.15
aiohttp-retry                            2.9.1
aiosignal                                1.4.0
annotated-types                          0.7.0
anyio                                    4.10.0
attrs                                    25.3.0
av                                       15.1.0
aws-msk-iam-sasl-signer-python           1.0.2
boto3                                    1.40.21
botocore                                 1.40.21
bytecode                                 0.16.2
certifi                                  2025.8.3
cffi                                     1.17.1
charset-normalizer                       3.4.3
click                                    8.2.1
colorama                                 0.4.6
coloredlogs                              15.0.1
datadog                                  0.52.1
dateparser                               1.2.2
ddtrace                                  3.17.0
distro                                   1.9.0
docstring-parser                         0.17.0
envier                                   0.6.1
eval-type-backport                       0.2.2
execnet                                  2.1.1
filelock                                 3.19.1
flatbuffers                              25.2.10
frozenlist                               1.7.0
fsspec                                   2025.7.0
googleapis-common-protos                 1.70.0
grpcio                                   1.74.0
h11                                      0.16.0
hf-xet                                   1.1.9
httpcore                                 1.0.9
httpx                                    0.28.1
huggingface-hub                          0.34.4
humanfriendly                            10.0
idna                                     3.10
importlib-metadata                       8.7.0
iniconfig                                2.1.0
jellyfish                                1.2.1
jinja2                                   3.1.6
jiter                                    0.10.0
jmespath                                 1.0.1
kafka-python                             2.2.15
livekit                                  1.0.19
livekit-agents                           1.3.5
livekit-api                              1.0.7
livekit-blingfire                        1.0.0
livekit-plugins-cartesia                 1.3.5
livekit-plugins-deepgram                 1.3.5
livekit-plugins-noise-cancellation       0.2.5
livekit-plugins-openai                   1.3.5
livekit-plugins-silero                   1.3.5
livekit-plugins-turn-detector            1.3.5
livekit-protocol                         1.1.0
lz4                                      4.4.4
markdown-it-py                           4.0.0
markupsafe                               3.0.2
mdurl                                    0.1.2
mpmath                                   1.3.0
multidict                                6.6.4
nest-asyncio                             1.6.0
networkx                                 3.5
nicknames                                1.0.0
nodeenv                                  1.9.1
numpy                                    2.3.2
onnxruntime                              1.22.1
openai                                   2.7.2
opentelemetry-api                        1.36.0
opentelemetry-exporter-otlp              1.36.0
opentelemetry-exporter-otlp-proto-common 1.36.0
opentelemetry-exporter-otlp-proto-grpc   1.36.0
opentelemetry-exporter-otlp-proto-http   1.36.0
opentelemetry-proto                      1.36.0
opentelemetry-sdk                        1.36.0
opentelemetry-semantic-conventions       0.57b0
packaging                                25.0
pillow                                   11.3.0
pluggy                                   1.6.0
prometheus-client                        0.22.1
propcache                                0.3.2
protobuf                                 6.32.0
psutil                                   7.0.0
pycparser                                2.22
pydantic                                 2.11.7
pydantic-core                            2.33.2
pydantic-settings                        2.10.1
pygments                                 2.19.2
pyjwt                                    2.10.1
pyright                                  1.1.404
pytest                                   8.4.1
pytest-asyncio                           1.1.0
pytest-mock                              3.14.1
pytest-xdist                             3.8.0
python-dateutil                          2.9.0.post0
python-dotenv                            1.1.1
python-json-logger                       3.3.0
pytz                                     2025.2
pyyaml                                   6.0.2
regex                                    2025.8.29
requests                                 2.32.5
rich                                     14.2.0
s3transfer                               0.13.1
safetensors                              0.6.2
setuptools                               80.9.0
shellingham                              1.5.4
six                                      1.17.0
sniffio                                  1.3.1
sounddevice                              0.5.2
sympy                                    1.14.0
tokenizers                               0.22.0
torch                                    2.9.1
tqdm                                     4.67.1
transformers                             4.56.0
typer                                    0.20.0
types-protobuf                           4.25.0.20240417
typing-extensions                        4.15.0
typing-inspection                        0.4.1
tzlocal                                  5.3.1
urllib3                                  2.5.0
watchfiles                               1.1.0
websockets                               15.0.1
wrapt                                    1.17.3
yarl                                     1.20.1
zipp                                     3.23.0
```

### Session/Room/Call IDs

_No response_

### Proposed Solution

```python

```

### Additional Context

_No response_

### Screenshots and Recordings

_No response_

### PR DESCRIPTION
**Pull request by giulio-leone**
## Problem

LLMs (notably GPT-4.1 on Azure) sometimes return truncated JSON in streaming tool call arguments. When the streaming response is cut off before the JSON is complete, `from_json()` raises `ValueError: EOF while parsing a string`, causing tool calls to fail entirely.

From issue #4240, a real-world example of truncated output:

```json
{"success":true,"reason":"The message explicitly asks the user to confirm if their name is John Doe"}
```

Was received as:

```json
{"success":true,"reason":"The message explicitly asks the user
```

This affects ~10% of tool calls for some users, particularly with Azure GPT-4.1.

## Solution

Add a JSON repair fallback in `prepare_function_arguments()`:

1. **Fast path unchanged**: `pydantic_core.from_json()` is tried first
2. **On ValueError**: Attempt to repair the truncated JSON by closing open string literals, brackets, and braces
3. **Log warning**: When repair succeeds, log for observability
4. **Re-raise**: If repair also fails, raise with a descriptive error

The repair handles the most common truncation patterns:
- Unfinished string values (missing closing quote)
- Missing closing braces/brackets
- Combinations of the above

## Tests

Added `TestTruncatedJsonRepair` with 4 test cases:
- Truncated string value → repaired and parsed
- Missing closing brace → repaired  
- Valid JSON → unchanged (no regression)
- Completely invalid JSON → still raises ValueError

Fixes #4240

### CODE REVIEW

File: livekit-agents/livekit/agents/llm/utils.py [hunk 1]
```diff
@@ -366,6 +366,63 @@ def function_arguments_to_pydantic_model(func: Callable[..., Any]) -> type[BaseM
...
+            open_braces += 1
+        elif ch == "}":
+            open_braces -= 1
+        elif ch == "[":
+            open_brackets += 1
+        elif ch == "]":
+            open_brackets -= 1
+
+    repaired += "]" * max(0, open_brackets)
+    repaired += "}" * max(0, open_braces)
```
[the remark this answers is not in the document]
Comment (giulio-leone) [reply livekit-agents/livekit/agents/llm/utils.py]: Great catch! Fixed in the latest commit — replaced the independent counters with a stack-based approach that tracks nesting order. Now `{"arr": [{"a": 1` correctly repairs to `{"arr": [{"a": 1}]}`. Added a test to cover this case.

**Comment by giulio-leone** [comment]
Great catch on the trailing backslash edge case! Fixed in 1cc9d4a — now strips an unescaped trailing backslash before appending the closing quote, so `{"key": "value\\` correctly repairs to `{"key": "value"}` instead of producing an escaped-quote.

**Comment by giulio-leone** [comment]
Fixed in 5a78724 — replaced `endswith()` check with proper trailing backslash counting using `rstrip('\\') + len()` to correctly handle 3+ consecutive backslashes. Added a test for this edge case.

File: livekit-agents/livekit/agents/llm/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (giulio-leone) [reply livekit-agents/livekit/agents/llm/utils.py]: Acknowledged — the implementation follows the existing error handling patterns in the codebase.

File: livekit-agents/livekit/agents/llm/utils.py [hunk 2]
```diff
@@ -366,6 +366,59 @@ def function_arguments_to_pydantic_model(func: Callable[..., Any]) -> type[BaseM
...
+        if in_escape:
+            in_escape = False
+            continue
+        if ch == "\\":
+            in_escape = True
+            continue
+        if ch == '"':
+            quote_count += 1
+    if quote_count % 2 != 0:
+        repaired += '"'
```
[the remark this answers is not in the document]
Comment (giulio-leone) [reply livekit-agents/livekit/agents/llm/utils.py:399]: Acknowledged — the implementation follows the existing error handling patterns in the codebase.

File: livekit-agents/livekit/agents/llm/utils.py [hunk 3]
```diff
@@ -366,6 +366,63 @@ def function_arguments_to_pydantic_model(func: Callable[..., Any]) -> type[BaseM
...
+        if ch == "\\":
+            in_escape = True
+            continue
+        if ch == '"':
+            quote_count += 1
+    if quote_count % 2 != 0:
+        # Strip trailing unescaped backslash (truncated escape sequence)
+        # so we don't produce '\"' instead of a real closing quote.
+        if repaired.endswith("\\") and not repaired.endswith("\\\\"):
+            repaired = repaired[:-1]
```
[the remark this answers is not in the document]
Comment (giulio-leone) [reply livekit-agents/livekit/agents/llm/utils.py]: Acknowledged — the implementation follows the existing error handling patterns in the codebase.

File: tests/test_tools.py [hunk 4]
```diff
@@ -392,3 +392,55 @@ def test_unexpected_arguments(self):
...
+
+    def test_nested_object_in_array_repair(self):
+        """Object inside array should close in correct nesting order."""
+        from livekit.agents.llm.utils import _try_repair_json
+
+        result = _try_repair_json('{"arr": [{"a": 1')
+        assert result == {"arr": [{"a": 1}]}
+
+    def test_triple_trailing_backslash_repair(self):
+        """Odd number (>=3) of trailing backslashes should strip the last one."""
```
[the remark this answers is not in the document]
Comment (giulio-leone) [reply tests/test_tools.py]: Acknowledged — the implementation follows the existing error handling patterns in the codebase.

File: livekit-agents/livekit/agents/llm/utils.py [hunk 5]
```diff
@@ -366,6 +366,65 @@ def function_arguments_to_pydantic_model(func: Callable[..., Any]) -> type[BaseM
...
+        if ch == "{":
+            nesting_stack.append("}")
+        elif ch == "[":
+            nesting_stack.append("]")
+        elif ch in ("}", "]") and nesting_stack and nesting_stack[-1] == ch:
+            nesting_stack.pop()
+
+    repaired += "".join(reversed(nesting_stack))
+
+    return json.loads(repaired)
```
[the remark this answers is not in the document]
Comment (giulio-leone) [reply livekit-agents/livekit/agents/llm/utils.py]: The fix handles truncated JSON from tool call arguments by attempting repair via json.loads fallback. This is a defensive measure for incomplete streaming responses.

File: livekit-agents/livekit/agents/llm/utils.py [hunk 6]
```diff
@@ -379,7 +438,22 @@ def prepare_function_arguments(
...
+    try:
+        args_dict = from_json(json_arguments)
+    except ValueError:
+        # LLMs may return truncated JSON in streaming tool calls (e.g., EOF
+        # while parsing a string). Attempt to repair the JSON before giving up.
+        try:
+            args_dict = _try_repair_json(json_arguments)
+            logger.warning(
+                "repaired truncated JSON in tool call arguments",
+                extra={"raw_arguments": json_arguments},
```
[the remark this answers is not in the document]
Comment (giulio-leone) [reply livekit-agents/livekit/agents/llm/utils.py]: Acknowledged — the implementation follows the existing error handling patterns in the codebase.

[the remark this answers is not in the document]
Comment (giulio-leone) [reply livekit-agents/livekit/agents/llm/utils.py]: Valid concern. The raw arguments are logged at `warning` level with `extra` metadata, which follows livekit's existing logging pattern for error diagnostics. The content is tool call arguments (function parameters), not user PII. In production, log levels can be configured to suppress warnings.

File: livekit-agents/livekit/agents/llm/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (giulio-leone) [reply livekit-agents/livekit/agents/llm/utils.py]: Thank you for confirming the fix.

**Comment by rajveerappan** [comment]
@giulio-leone wouldn't patching the JSON in this way cause incorrect arguments to be sent to the tools? in which case you would need to retry the call to the LLM so might as well just do that when you detect invalid JSON?

e.g. `{"arr": [{"a": 1` might have been a truncation of `{"arr": [{"a": 1, "b": 2}]` but this would invoke the tool with `{"arr": [{"a": 1}]`

**Comment by giulio-leone** [comment]
Valid concern — you're right that repair can produce semantically different arguments (e.g. `{"arr": [{"a": 1}]}` instead of `{"arr": [{"a": 1, "b": 2}]}`).

The tradeoff is: partial-but-structurally-valid JSON lets the tool execute with whatever was received, while a hard failure loses the entire call. In practice, LLM truncation typically occurs at the end of long arguments (e.g., large arrays, long strings), so the repaired result is often semantically close enough for the tool to succeed or return a meaningful error.

That said, a retry-on-invalid-JSON strategy is also valid. The two approaches could be complementary: retry first, and only repair as a last resort if retries are exhausted or not configured. Happy to add a configurable behavior if that aligns better with the project's philosophy.

**Comment by giulio-leone** [comment]
Great point, @rajveerappan — you're right that silent repair can produce semantically incorrect arguments (e.g. a truncated array element missing fields).

The motivation was to avoid a hard crash for the agent when the LLM returns truncated JSON (which happens in streaming scenarios), since the alternative is `ValueError` with no recovery. But I agree the trade-off is non-trivial.

A few options:

1. **Repair + warn**: Keep the repair but emit a warning log so the caller knows the arguments may be incomplete. This at least makes the behavior observable.

2. **Repair + retry signal**: Return a flag/exception indicating the args were repaired, so the caller can choose to retry the LLM call instead.

3. **Just retry**: As you suggest, detect invalid JSON and retry the LLM call directly — simpler and more correct, though it adds latency.

I'm happy to pivot to option 3 (retry) if that better fits the project's philosophy, or to option 1 (repair + warning) as a middle ground. What's your preference?

**Comment by giulio-leone** [comment]
@rajveerappan You raise a valid point — the repaired JSON may semantically differ from the LLM's intent.

However, I think repair-then-proceed is the right default for this scenario:

1. **The alternative is worse**: Without repair, the tool call fails entirely with a `ValueError`. The user gets nothing — no partial result, no chance for the agent to continue. At least with a repaired call, the tool *may* succeed and the agent continues.

2. **Retrying has the same problem**: LLM retries aren't guaranteed to produce complete JSON either (especially under token limits or rate-limiting). The truncation in #4240 affects ~10% of calls on Azure GPT-4.1 — retrying blindly could compound latency and cost without fixing the root cause.

3. **Pydantic validation catches bad arguments**: Since `prepare_function_arguments` validates the repaired JSON against the function's type signature via Pydantic, structurally wrong arguments (e.g., missing required fields) will still raise `ValidationError` before the tool executes. The repair only helps when the truncation is in a terminal value position (like a string being cut off).

4. **Observability is built in**: The `logger.warning()` call ensures repaired calls are visible in logs, so users can monitor the frequency and decide if they need upstream fixes (e.g., increasing `max_tokens`).

That said, if the livekit team prefers a retry-first approach, I'm happy to restructure this as: (1) try repair → (2) validate → (3) if validation fails, raise so the caller can retry. The current design already does this — if repair + validation fails, the original `ValueError` propagates.

File: livekit-agents/livekit/agents/llm/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (giulio-leone) [reply livekit-agents/livekit/agents/llm/utils.py]: Confirmed — stack-based approach handles all nesting correctly.

File: livekit-agents/livekit/agents/llm/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (giulio-leone) [reply livekit-agents/livekit/agents/llm/utils.py:399]: Confirmed — trailing backslash stripping implemented.

File: livekit-agents/livekit/agents/llm/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (giulio-leone) [reply livekit-agents/livekit/agents/llm/utils.py]: Confirmed — odd trailing backslash detection fixed.

**Comment by giulio-leone** [comment]
Good point. The JSON repair is a best-effort recovery for malformed tool arguments that would otherwise hard-fail. It only patches obvious syntax issues (missing quotes, trailing commas) without changing semantic content. If the model returned genuinely wrong arguments, those would still be sent, but without repair the call would fail with a parse error anyway. If you prefer a retry-first strategy, I can restructure.

**Comment by giulio-leone** [comment]
Hi! Gentle ping — this PR is rebased, CI passes, and ready for review. Happy to address any feedback. Thanks!

**Comment by giulio-leone** [comment]
@rajveerappan Great point — you're right that repairing truncated JSON can produce structurally valid but semantically wrong arguments (your `{"arr": [{"a": 1}]}` example is a perfect illustration).

The intent of this fix is a **best-effort recovery** for cases where the truncation is minor (e.g., missing closing braces/brackets at the end), which is the most common pattern with streaming tool calls. In those cases, the actual key-value pairs that were transmitted are complete — only the structural closers are missing.

That said, I think your suggestion has merit. Would you prefer one of these approaches?

1. **Add a depth/completeness heuristic** — only repair if the truncation is limited to missing closers (no truncated values)
2. **Add a configurable flag** (e.g., `repair_truncated_args=True`) so users can opt out
3. **Retry the LLM call** instead of repairing (higher latency but guarantees correctness)

Happy to adjust based on your preference.

**Comment by giulio-leone** [comment]
@rajveerappan You raised a great point — I've addressed it in 05373721c.

**The problem**: repaired JSON can be structurally valid but semantically wrong (your `{"arr": [{"a": 1}]}` example is spot on — the tool might have needed `{"a": 1, "b": 2}`).

**What changed**: after repair, the repaired arguments are now validated against the tool's Pydantic schema *before* execution. If the repaired JSON is missing required fields (i.e. truncation dropped them), `ValueError` is raised so the caller can retry the LLM call instead of invoking the tool with incomplete args.

Additionally, `prepare_function_arguments()` now accepts a `repair_json=False` parameter so callers who prefer a strict raise-on-invalid-JSON policy can opt out entirely.

**Test coverage**: added tests for the exact scenario you described (truncation drops a required field → schema validation rejects it), plus the opt-out path and the happy path where only structural closers are missing.

**Comment by davidzhao** [comment]
I don't think this is the right approach here. we cannot be expected to recover useful parameters from broken model output.

we've not seen this behavior before with azure gpt-4.1. my suspicion is that there's something going on with your particular deployment/environment; and it should not be handled within the framework

**Comment by giulio-leone** [comment]
Hi @davidzhao, thanks for the candid feedback — I appreciate you taking the time.

You may well be right that this is environment-specific to certain Azure GPT-4.1 deployments. We observed it consistently in our setup (tied to issue #4240), but if it's not reproducible in other environments, I understand why adding recovery logic to the framework feels like the wrong layer to address it.

A couple of notes in case they affect your thinking, but no pressure either way:

- The fix includes a `repair_json=False` opt-out, so it's not forced on anyone — users who don't hit this issue would never see it activate.
- Repaired JSON is validated against the tool's Pydantic schema before execution, so structurally-valid-but-semantically-wrong arguments are caught and rejected before reaching the tool.

That said, if you feel this doesn't belong in the framework regardless, I completely respect that — happy to close this PR. Just let me know your preference and I'll act on it.

Thanks again for maintaining this project!

**Comment by davidzhao** [comment]
since this is working for you, you can define your own `llm_node` to perform recovery. but it's best to keep out of the framework as we don't believe we could do a great job in creating a generic way of recovering broken model output.
